src/pyfmi/fmi3.pyx

```diff
@@ -675,6 +678,155 @@ cdef class FMUModelBase3(FMI_BASE.ModelBase):
+        cdef FMIL3.fmi3_import_variable_list_t*   variable_list
+        cdef FMIL.size_t                          variable_list_size
+        cdef FMIL3.fmi3_value_reference_t         value_ref
+        cdef FMIL3.fmi3_base_type_enu_t           data_type
+        cdef FMIL3.fmi3_variability_enu_t         data_variability
+        cdef FMIL3.fmi3_causality_enu_t           data_causality
+        cdef FMIL3.fmi3_initial_enu_t             data_initial
+
+        cdef int has_start = 0
+        # TODO: Can we rename the keyword arg 'filter' to variable_filter?
+        variable_filter = filter
+        # TODO: Can we rename the keyword arg 'type' to variable_type?
+        variable_type = type
+        cdef list filter_list, variable_return_list = []
+        variable_dict = {}
+
+        variable_list      = FMIL3.fmi3_import_get_variable_list(self._fmu, 0)
+        variable_list_size = FMIL3.fmi3_import_get_variable_list_size(variable_list)
+
+        """
+            TODO: If we decide to force user to use FMI3_* enums as input we need to also error check
+                  such that user doesn't specify for example causality = FMI3_Type.X or similar.
+                  Also let's remove '_as_list'?
+        """
+
+        for index in range(variable_list_size):
+            variable = FMIL3.fmi3_import_get_variable(variable_list, index)
+
+            has_start = FMIL3.fmi3_import_get_variable_has_start(variable)
+            if only_start and has_start == 0:
+                continue
+
+            data_variability = FMIL3.fmi3_import_get_variable_variability(variable)
+
+            if only_fixed:
+                # fixed variability requires start-value
+                if has_start == 0:
+                    continue
+                elif data_variability != FMI3_Variability.FIXED.value:
+                    continue
+
+            name             = pyfmi_util.decode(FMIL3.fmi3_import_get_variable_name(variable))
+            value_ref        = FMIL3.fmi3_import_get_variable_vr(variable)
+            description      = self._get_variable_description(variable)
+            data_causality   = FMIL3.fmi3_import_get_variable_causality(variable)
+            data_initial     = FMIL3.fmi3_import_get_variable_initial(variable)
+            data_type        = FMIL3.fmi3_import_get_variable_base_type(variable)
+            #If only variables with start are wanted, check if the variable has start
+
+            # TODO: Discuss if we want to support also regular integers as inputs
+            if isinstance(variable_type, FMI3_Type) and (data_type != variable_type.value):
+                continue
+            if isinstance(variability, FMI3_Variability) and (data_variability != variability.value):
+                continue
+            if isinstance(causality, FMI3_Causality) and (data_causality != causality.value):
+                continue
+
+            if variable_filter:
+                filter_list = self._convert_filter(variable_filter)
```

Comment:
I think this absolutely needs to be moved out of the loop to avoid converting the filter list for each variable.

Reply:
Good catch, fixed!